segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,483 @@
+            for (String systemSegment : systemSegments) {
+                SegmentMetadata segmentMetadata = SegmentMetadata.builder()
+                        .name(systemSegment)
+                        .ownerEpoch(epoch)
+                        .maxRollinglength(segmentRollingPolicy.getMaxLength())
+                        .build();
+                segmentMetadata.setActive(true).setOwnershipChanged(true);
+                segmentMetadata.checkInvariants();
+                txn.create(segmentMetadata);
+                txn.markPinned(segmentMetadata);
+            }
+
+            // Step 2: For each epoch, find the corresponding system journal files, process them and apply operations recorded.
+            applySystemLogOperations(txn, chunkStartOffsets, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
+
+            // Step 3: Adjust the length of the last chunk.
+            adjustLastChunkLengths(txn);
+
+            // Step 4: Apply the truncate offsets.
+            applyFinalTruncateOffsets(txn, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
+
+            // Step 5: Finally commit all data.
+            txn.commit(true, true);
+        }
+    }
+
+    /**
+     * Process all systemLog entries to recreate the state of metadata storage system segments.
+     */
+    private void applySystemLogOperations(MetadataTransaction txn, HashMap<String, Long> chunkStartOffsets, HashMap<String, Long> finalTruncateOffsets, HashMap<String, Long> finalFirstChunkStartsAtOffsets) throws ChunkStorageException, IOException, StorageMetadataException {
+        for (int epochToRecover = 0; epochToRecover < epoch; epochToRecover++) {
+            // Start scan with file index 0.
+            int fileIndexToRecover = 0;
+            while (chunkStorage.exists(getSystemJournalFileName(containerId, epochToRecover, fileIndexToRecover))) {
+                val systemLogName = getSystemJournalFileName(containerId, epochToRecover, fileIndexToRecover);
+                val info = chunkStorage.getInfo(systemLogName);
+                val h = chunkStorage.openRead(systemLogName);
+                byte[] contents = new byte[Math.toIntExact(info.getLength())];
+                long fromOffset = 0;
+                int remaining = contents.length;
+                while (remaining > 0) {
+                    int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset));
+                    remaining -= bytesRead;
+                    fromOffset += bytesRead;
+                }
+                BufferedReader reader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(contents)));
+                String line;
+
+                while ((line = reader.readLine()) != null) {
+                    // Handle only whole records.
+                    if (line.startsWith(START_TOKEN) && line.endsWith(END_TOKEN)) {
+                        String[] records = line.split(RECORD_SEPARATOR);
```

Comment:
IT is more efficient to use a tokenizer which returns an iterator vs doing split.

Reply:
At the current moment I need debug friendly format.
Once I am done with all changes, I do plan to implement serialization at the very end as part of the last PR in this group.